typewriter/fixes/fixer_utils.py

```diff
@@ -133,3 +140,254 @@ def touch_import(package, name, node):
+
+
+def decompose_name(node):
+    """
+    NOTE: Per the lib2to3 grammar:
+            dotted_name: NAME ('.' NAME)*
+          This means that dotted_name can be either dotted or not dotted, i.e. it's a generalized
+          form of NAME. So this function will cover both cases.
+
+    Given a dotted_name node this will return a tuple of the form (pkg, name, full_string) where
+    all are str
+        ex: a.b.c => (a.b, c, a.b.c)
+            b.c => (b, c, b.c)
+            c => (None, c, c)
+    otherwise it will return None for each field
+    """
+    if node.type == token.NAME:
+        # node is just a name, no dots
+        return None, node.value, node.value
+
+    if node.children:
+        # Right most node will be the name, i.e. a.b.c = ['a','.','b','.','c']
+        name_node = node.children[-1]
+        package_nodes = node.children[:-2]
+        name = str(name_node).strip()
+        package = ''.join(str(n).strip() for n in package_nodes)
+        full = ''.join(str(n).strip() for n in node.children)
+        return package, name, full
+
+    return None, None, None
+
+
+def get_import_info(node):
+    """
+    Wraps the cachable get_import_info in order to convert the node into a hashable node before
+    attempting to cache
+    """
+    node = HashableNode(node)
+    return _get_import_info_cachable(node)
+
+
+class HashableNode(object):
+    """
+    Hashable wrapper for node and leaf objects in order to cache
+    """
+
+    def __init__(self, node):
+        # type: (Union[Node, Leaf]) -> None
+        """
+        Parameters
+        -----------
+        node : Union[Node, Leaf]
+        """
+        self.node = node
+
+    def __hash__(self):
+        return hash((self.node.get_lineno(),
+                     self.node.depth(),
+                     self.node.__str__()))
+
+
+@cache
+def _get_import_info_cachable(hashable_node):
+    # type: (HashableNode) -> ImportInfo
+    """
+    We cache using hashable_node as a key and analyze hashable_node.node
+
+    If node is a valid import_stmt, this will return a named tuple for each component of the
+    statement
+        ex: from a.b import c as d, e  => (imports=[(import_name='c',
+                                                     binding_name='d'),
+                                                    (import_name='e',
+                                                     binding_name='e')],
+                                           package='a.b',
+                                           node=node)
+
+    Since we will regularly iterate over the list import node for their information when
+    searching for binding or import matches, we cache the results of this function for
+    quick and non-redundant lookups
+
+    Parameters
+    -----------
+    hashable_node : HashableNode
+
+    Return
+    -----------
+    ImportInfo
+    """
+    # Get the node from the hashable_node wrapper
+    node = hashable_node.node
+
+    def handle_name(node, allow_dotted=True):
```

Comment:
I don't see the need for the `allow_dotted` kwarg - this should just return the "correct" result, regardless of what it's fed.

Currently, `allow_dotted` is set to False if we have an `import_as_name`, but in that case, we should never get a `dotted_as_name` anyway, due to grammar rules.  Also, `decompose_name` should be able to handle dotted and non-dotted names just fine.